### ISSUE TITLE
Tracking Issue: Prepare the private version of the Interactivity API for WP 6.4

### ISSUE THREAD

**Issue by luisherranz** [WordPress/gutenberg#54291]
WP 6.4 will include Gutenberg up to the 16.7 version, which includes the new Image Lightbox, an enhanced pagination for the Query block and accessibility fixes for the Navigation block, all of them powered by the Interactivity API. Although the Interactivity API is still in the _proposal phase_, we can include a private version that only Core blocks can access in WP 6.4 to release these features/enhancements. 

Please note that **the Interactivity API is still likely to change** before the final public version is released, but by keeping it private to Core blocks in WP 6.4, we can still modify it before it's public. Furthermore, if the proposal is finally rejected for some reason, we just have to replace the Interactivity API implementation of the Core blocks with whatever new approach is adopted instead.

The release of the 16.7 RC is scheduled for the 20th of September. Before that date, we need to complete the following tasks:

- [x] Query block
  - [x] Only allow enabling if all blocks inside Post Template are Core blocks to make sure they are compatible
    - https://github.com/WordPress/gutenberg/pull/54347
  - [x] Remove the usage of `wp_store`
    - https://github.com/WordPress/gutenberg/pull/54359
  - [x] Add timeout to client navigation
     - https://github.com/WordPress/gutenberg/pull/54474
  - [x] Additional design enhancements
    - https://github.com/WordPress/gutenberg/issues/54414
    - https://github.com/WordPress/gutenberg/pull/54455
  - Low Priority
    - [x] Improve prefetching strategy to prefetch next/previous links once the user has clicked once on them.
      - https://github.com/WordPress/gutenberg/pull/54781
    - [ ] Add a fading animation
      - https://github.com/WordPress/gutenberg/issues/54060
      - https://github.com/WordPress/gutenberg/pull/54746
    - [ ] ~~Figure out if we can combine the focus scroll with the smooth scroll of scrollIntoView.~~ It only works in Chrome.
      - https://github.com/WordPress/gutenberg/pull/54746
    - [ ] ~~Add a “blink” the Query block in the Editor as soon as the toggle is enabled, just to provide some visual feedback in the canvas that something changed.~~ Discarded for now.
- [x] Navigation block
  - [x] Remove `gutenberg_should_block_use_interactivity_api` and remove old implementation
    - https://github.com/WordPress/gutenberg/pull/54297
  - [x] Make sure HTML is fine without SSR
    - https://github.com/WordPress/gutenberg/pull/54343
- [x] File block 
  - [x] Remove `gutenberg_should_block_use_interactivity_api` and remove old implementation
    - https://github.com/WordPress/gutenberg/pull/54297
  - [x] Make sure HTML is fine without SSR
    - https://github.com/WordPress/gutenberg/pull/54343
- [x] Search block
  - [x] Finish the new implementation and remove the old one and make sure HTML is fine without SSR
    - https://github.com/WordPress/gutenberg/pull/53343
- [x] Image Lightbox
  - [x] Make sure HTML is fine without SSR and rename `view-interactivity.js` file to `view.js`
    - https://github.com/WordPress/gutenberg/pull/54547
  - [x] Remove the use of `WP_Theme_JSON_Resolver_Gutenberg`
    - https://github.com/WordPress/gutenberg/pull/53851
- [x] Prepare WP Core to switch package from public to private
  - [x] Modify the Webpack config to bundle `packages/interactivity` privately and enqueue the new files
    - https://github.com/WordPress/gutenberg/pull/54311
  - [x] Finish PR in WP Core to merge it once the 16.7 RC is merged
    - https://github.com/WordPress/wordpress-develop/pull/5195
  - [x] Ensure that the `wp-interactivity` dependency is always added
    - https://github.com/WordPress/gutenberg/pull/54594

**Comment by SantosGuillamot** [comment]
I created this pull request to remove `gutenberg_should_block_use_interactivity_api` and the old implementation of the Navigation and File blocks. In the end, I decided to do everything simultaneously because of some related conditionals.

https://github.com/WordPress/gutenberg/pull/54297

**Comment by luisherranz** [comment]
Approved! Thanks, Mario 👍

**Comment by luisherranz** [comment]
I've opened a PR to test the WP Core webpack configuration. It's not finished yet, but so far it seems to be working great:

- https://github.com/WordPress/gutenberg/pull/54311

**Comment by SantosGuillamot** [comment]
I've opened a PR to add manual SSR to the Navigation block and the File block. As explained in the pull request, I believe that in the File block, it is complicated to replicate the JS logic, so we have to make a decision there.

* https://github.com/WordPress/gutenberg/pull/54343

**Comment by SantosGuillamot** [comment]
I've been reviewing the Search block PR, and I think it should be ready to be merged, it needs review and approval. I also modified it so it works without the Interactivity API SSR.

* https://github.com/WordPress/gutenberg/pull/53343

**Comment by DAreRodz** [comment]
I've opened a draft PR for disabling the enhanced pagination if third-party blocks exist in the Post Template.

- https://github.com/WordPress/gutenberg/pull/54347

**Comment by luisherranz** [comment]
I've added a new task for the Query block:

- Remove the usage of `wp_store`.

**Comment by SantosGuillamot** [comment]
I've created a pull request to remove the usage of `wp_store` in the Query block:

* https://github.com/WordPress/gutenberg/pull/54359

**Comment by SantosGuillamot** [comment]
The PR to move the Search block to the Interactivity API has been merged:

* https://github.com/WordPress/gutenberg/pull/53343

**Comment by luisherranz** [comment]
I've sucessfully finished modifying the Webpack files to make `@wordpress/interactivity` a private package:

- https://github.com/WordPress/gutenberg/pull/54311#issuecomment-1716033941

The next step is to try to test it out in WP Core. I've also added a couple of related tasks to the opening post.

**Comment by luisherranz** [comment]
I've ported the changes to WordPress Core and they are working great. This should be ready to merge once the sync between Gutenberg and WP Core happens (with the correct modifications to the PR).

- https://github.com/WordPress/wordpress-develop/pull/5195

**Comment by luisherranz** [comment]
@artemiomorales, @michalczaplinski: @tellthemachines mentioned me that we need to review the usage of `WP_Theme_JSON_Resolver_Gutenberg` in the Image block before the 16.7 RC release (20th of Sept).

This is the exact message:

> We're calling `WP_Theme_JSON_Resolver_Gutenberg` in the [Image block PHP file](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/index.php#L41), which isn't defined in core. We'll need to add some logic similar to [this](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/index.php#L315C2-L315C2) to check if the core class exists.

I've added to the list of tasks in the opening post. Please take care of it and mark it as ready afterward. Thanks 🙂

**Comment by luisherranz** [comment]
I've opened an issue with some design enhancements mentioned by @jasmussen. There are some that we should do before the 16.7 RC release, and others that we can try if we have time:

- https://github.com/WordPress/gutenberg/issues/54414

**Comment by DAreRodz** [comment]
The changes to allow only core blocks inside a Query Loop with enhanced pagination are done and merged.

- https://github.com/WordPress/gutenberg/pull/54347

I'll continue with the rest of the enhancements.

**Comment by DAreRodz** [comment]
This addresses most of the [design feedback](https://github.com/WordPress/gutenberg/issues/54414) from @jasmussen. 👇

- https://github.com/WordPress/gutenberg/pull/54455

It's ready for review.

**Comment by DAreRodz** [comment]
I've created another PR that adds a `timeout` option to `navigate()`, with a default value of `10000` milliseconds (10 seconds).

- https://github.com/WordPress/gutenberg/pull/54474

**Comment by luisherranz** [comment]
David opened another PR to improve the accessibility of the modal:

- https://github.com/WordPress/gutenberg/pull/54507

**Comment by luisherranz** [comment]
> Open PR in Gutenberg to [add this code](https://github.com/WordPress/gutenberg/pull/54311#issuecomment-1716033941) to each interactive block to make sure that all the scripts have the `wp-interactivity` dependency

~~I think it's safe to do this in WordPress itself. I'll add it to this PR instead:~~ (see below)

**Comment by michalczaplinski** [comment]
> We're calling WP_Theme_JSON_Resolver_Gutenberg in the [Image block PHP file](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/index.php#L41), which isn't defined in core. We'll need to add some logic similar to [this](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/index.php#L315C2-L315C2) to check if the core class exists.

This issue was fixed by merging https://github.com/WordPress/gutenberg/pull/53851 which removes the reference to `WP_Theme_JSON_Resolver_Gutenberg` in the Image block. There is another reference to `WP_Theme_JSON_Resolver_Gutenberg` in https://github.com/WordPress/gutenberg/blob/62c1f50704d7b2235a525e121d5ee6ea0bc449e1/lib/block-supports/behaviors.php#L67 which is temporarily added to Gutenberg for backward-compatibility and will be removed in GB 17.0.

**Comment by SantosGuillamot** [comment]
I've created a PR to handle the remaining tasks related to the lightbox:

* https://github.com/WordPress/gutenberg/pull/54547

**Comment by tellthemachines** [comment]
> > Open PR in Gutenberg to [add this code](https://github.com/WordPress/gutenberg/pull/54311#issuecomment-1716033941) to each interactive block to make sure that all the scripts have the `wp-interactivity` dependency
> 
> I think it's safe to do this in WordPress itself. I'll add it to this PR instead:

@luisherranz does this require changes to block PHP files? If so, it has to be done in Gutenberg. The block PHP gets generated in core from the `block-library` package so it can't be manually edited there.

**Comment by luisherranz** [comment]
> @luisherranz does this require changes to block PHP files?

No, it doesn't. ~~We can add it in Core in `wp-includes/script-loader.php` if that's ok.~~

EDIT: I tried, but I think it's simpler to add it here in Gutenberg:

- https://github.com/WordPress/gutenberg/pull/54594

**Comment by luisherranz** [comment]
I added logic to start prefetching the next pages on the first click to the next and previous buttons:

- https://github.com/WordPress/gutenberg/pull/54781

**Comment by DAreRodz** [comment]
I opened a PR that shows a subtle animation when the HTML is updated and changes the scroll behavior to `smooth`  during navigation. Ready for review.

- https://github.com/WordPress/gutenberg/pull/54746

**Comment by ermincelikovic** [comment]
Hi folks, thanks for the good work!

I just wanted to understand better - Interactivity API is not going to be useable outside of WP Core in 6.4?
Does it mean it is planned for 6.5?

**Comment by SantosGuillamot** [comment]
> I just wanted to understand better - Interactivity API is not going to be useable outside of WP Core in 6.4?
Does it mean it is planned for 6.5?

Yes, in 6.4, only core blocks can use the Interactivity API. If you want to use the Interactivity API for your custom blocks, you need to install the Gutenberg plugin as well. We considered it safer this way because there are some parts that need more testing and the API is likely to change. So it might require developers who built custom blocks using the API to change the syntax.

On the other hand, if everything goes as expected, the idea is to make it public in WP Core in 6.5 as you say.

**Comment by luisherranz** [comment]
Even though there's still some [discussion about the fading animation of the Query block](https://github.com/WordPress/gutenberg/pull/54746), I'm going to close this Tracking Issue as finished because WP 6.4 Beta 1 has already been released.

We will continue working in the new `store()` API, please follow the progress in this Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/53740
